giraph-core/src/main/java/org/apache/giraph/comm/flow_control/CreditBasedFlowControl.java

```diff
@@ -215,10 +231,38 @@ public void run() {
         }
       }
     });
-    thread.setUncaughtExceptionHandler(exceptionHandler);
-    thread.setName("resume-sender");
-    thread.setDaemon(true);
-    thread.start();
+    resumeHandlerThread.setUncaughtExceptionHandler(exceptionHandler);
+    resumeHandlerThread.setName("resume-sender");
+    resumeHandlerThread.setDaemon(true);
+    resumeHandlerThread.start();
+
+    // Thread to handle/send cached requests
+    Thread cachedRequestHandlerThread = new Thread(new Runnable() {
+      @Override
+      public void run() {
+        while (true) {
+          Pair<Integer, WritableRequest> pair = null;
+          try {
+            pair = toBeSent.take();
+          } catch (InterruptedException e) {
+            throw new IllegalStateException("run: failed while waiting to " +
+                "take an element from the request queue!", e);
+          }
+          int taskId = pair.getLeft();
+          WritableRequest request = pair.getRight();
+          nettyClient.doSend(taskId, request);
+          if (aggregateUnsentRequests.decrementAndGet() == 0) {
+            synchronized (aggregateUnsentRequests) {
+              aggregateUnsentRequests.notifyAll();
+            }
+          }
+        }
+      }
+    });
+    cachedRequestHandlerThread.setUncaughtExceptionHandler(exceptionHandler);
+    cachedRequestHandlerThread.setName("cached-req-sender");
+    cachedRequestHandlerThread.setDaemon(true);
+    cachedRequestHandlerThread.start();
```

Comment:
You can create a utility method like ThreadUtils.startThread with exception handler.